src/historywork/BatchWork.h

```diff
@@ -0,0 +1,40 @@
+// Copyright 2018 Stellar Development Foundation and contributors. Licensed
+// under the Apache License, Version 2.0. See the COPYING file at the root
+// of this distribution or at http://www.apache.org/licenses/LICENSE-2.0
+#pragma once
+
+#include "work/Work.h"
+#include <medida/meter.h>
+#include <medida/metrics_registry.h>
+
+namespace medida
+{
+class Meter;
+}
+
+namespace stellar
+{
+
+class TmpDir;
+
+class BatchWork : public Work
+{
+    /* This class performs parallel batching of Work by throttling workers.
+       Child classes must supply iteration methods, that would generate work
+       they'd like to perform. This class only acts as a commander, adding more
+       work if it has bandwidth.
+    **/
+  public:
+    BatchWork(Application& app, WorkParent& parent, std::string name);
+    ~BatchWork() override;
+    void onReset() override;
+    void notify(std::string const& child) override;
+
+    // Optionally mark metrics
+    virtual void markMetrics(Work& work){};
```

Comment:
`notify` logically belongs to BatchWork and is the same for all children (with minor variation in metrics), so I did not want to duplicate same code in child classes. I also did not want for children to override `notify` with a dependency on parent's `notify`.

Reply:
I am not sure I understand what you are saying: one of the typical patterns for overriding a method is to call the parent implementation to avoid duplication. I don't see why metrics should be treated differently in the context of `batchWork`, maybe there should be a `updateMetrics` member (that doesn't take parameters) exposed at the `Work` level, but that seems to be a different conversation.

Reply:
Although common, I don't think it's necessarily right here, as it requires every subclass to know that it must call parent's notify. 
I was also thinking if it makes sense to move metrics into download work classes (like GetAndUnzipRemoteFileWork) instead of batching classes. That would require keeping track of file types though.